part of the infrastructure for public & private dependencies in the resolver

Eh2406 opened the pull request:
This is part of my work on public & private dependencies in the resolver from #6129. As discussed there the proptest fuzzers are happy to find exponential blow up with all the back jumping strategies I have tried. So this PR does not have a back jumping strategie nor does it have the proptest fuzzers generating public dependencies. These will both need to change for the feature to stabilize. In the meantime it gives the correct results on the cases it can handle.

With https://github.com/rust-lang/rust/pull/57586 landed there is a lot of work to do on Cargos front end. Adding a UI for this, passing the relevant things to rustc, passing it to crates.io, passing it to cargo-metadata. This is good enough to allow that work to proceed.

rust-highfive commented:
r? @dwijnand

(rust_highfive has picked a reviewer for you, use r? to override)

dwijnand commented:
r? @alexcrichton

Eh2406 commented:
Push a fix for my braking every test.

alexcrichton commented on src/cargo/core/resolver/context.rs:
```diff
@@ -26,6 +26,12 @@ pub struct Context {
     pub resolve_features: im_rc::HashMap<PackageId, Rc<HashSet<InternedString>>>,
     pub links: im_rc::HashMap<InternedString, PackageId>,
+    pub public_dependency:
+        im_rc::HashMap<PackageId, im_rc::HashMap<InternedString, (PackageId, bool)>>,
```
We should probably already have this for all the other fields but could a comment be added here for this new field about what this is a map to/from?

alexcrichton commented on src/cargo/core/resolver/context.rs:
```diff
@@ -26,6 +26,12 @@ pub struct Context {
+
+    // This is somewhat redundant with the `resolve_graph` that stores the same data,
+    //   but for querying in the opposite order.
+    pub parents: Graph<PackageId, Rc<Vec<Dependency>>>,
```
In our past debugging we've found that cloning a `Context` happens so often that an expensive clone means a slower resolver, so I was curious about this! 

I think we explicitly removed `Graph` from this data structure earlier in favor of `RcList` below, so was curious on your thoughts on this.

I haven't read the whole patch yet though, so I'll likely find what happens soon.

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -591,6 +596,44 @@ fn activate(
+            .iter()
+            .flat_map(|x| x.values())
+            .filter_map(|x| if x.1 { Some(&x.0) } else { None })
+            .chain(Some(candidate.summary.package_id()).iter())
```
The trailing `.iter()` here I think can be removed

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -591,6 +596,44 @@ fn activate(
             candidate.summary.package_id(),
             dep.clone(),
         ));
+        Rc::make_mut(
```
Could a comment be added here to explain what's going on?

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -725,6 +769,30 @@ impl RemainingCandidates {
                     continue;
                 }
             }
+            for &t in cx
```
Could a comment be added here abot what candidates this is rejecting?

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -799,6 +868,9 @@ fn find_candidate(
         if !backtracked
+            && !conflicting_activations
+                .values()
+                .any(|c| *c == ConflictReason::PublicDependency)
```
Could the comment be updated to account for this new clause?

alexcrichton commented on src/cargo/core/resolver/types.rs:
```diff
@@ -420,6 +425,8 @@ impl ConflictReason {
     }
 }
 
+pub type Conflict = BTreeMap<PackageId, ConflictReason>;
```
s/Conflict/ConflictMap/ perhaps?

alexcrichton commented on src/cargo/util/graph.rs:
```diff
@@ -1,35 +1,36 @@
+use im_rc;
+
+pub struct Graph<N: Clone, E: Clone> {
+    nodes: im_rc::OrdMap<N, im_rc::OrdMap<N, E>>,
```
aha, I see how this would make things a bit speedier on the cloning half of things

alexcrichton replied:
(maybe `RcList` is now obsoleted?)

Eh2406 commented:
I added comments, I think it addresses your concerns.

The test point out that the RFC did not discuss how this interacts with the other things cargo can do. I don't have a mental model of how this should work.

Eh2406 commented:
So the specific questions are:

How does this interact with renamed dependencies?
--------
There are lots of juicy corner cases here. On one had being able to depend on 2 different versions of the same library is one of the points of renamed dependencies. Fore example so you can publicly derive `serde1` and `serde2` traits on the same type. On the other hand making sure you and your dependencies agree on the same version of `serde` is the hole point of public dependencies.

This makes my head hurt, is there something simple I am missing? If not, where is the appropriate place to discuss fundamental problems with an accepted RFC?

How does this interact with `'cfg({})'`ing in different versions?
----------
Conceptually this one is very simple. Instead of asking "is there a package that can see two different versions of a dependency" we need to ask "is there a set of arguments so that there is a package that can see two different versions of a dependency". But the complexity is in the details. If we are not careful this is a SAT problem that needs to be solved for each tick. Where is the existing `'cfg({})'` logic handled?

Is there some minimal subset of this feature that we can get to merge while we work the issues out? Like how this PR does not include backtracking nor error messages.

Eh2406 commented:
I got test to pass by just making deps that have a cfg or a rename invisible. This is not a sablizable solution, but it does get the test to pass.

bors commented:
:umbrella: The latest upstream changes (presumably #6687) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -585,35 +590,80 @@ fn activate(
+        // other functionality is used.
+        // TODO: this disable is not a long term solution.
+        //       This needs to be removed before public dependencies are stabilized!
+        if dep.platform().is_none() && dep.explicit_name_in_toml().is_none() {
```
I think the behavior that we want is to probably just delete this `if` clause and unconditionally run the below?

Eh2406 replied:
Yes, we would like to delete this if statement and have the code below deal with the special handling of these corner cases. At the beginning of this PR we did not have this if statement and it broke the existing tests. Until we figure out how the corner cases are supposed to be handled, this if statement lets other people make progress.

alexcrichton replied:
Er sorry it's taking me awhile to get back to this, but do you have a corner case in mind that this is needed for? I can't actually think of a case where the platform and/or renaming would cause issues

Eh2406 replied:
It is easy to find the examples, just look at the test that do not pass without it.
- renaming: [this test](https://github.com/rust-lang/cargo/blob/d75d1fbe6b43e59d0a04e73659e411855c2633fc/tests/testsuite/rename_deps.rs#L7) means that `foo` can see `bar 0.1.0` and `bar 0.2.0`. `foo` can get `expected bar::Thing but got bar::Thing` errores so is not permitted by public & private dependencies.
- platform: [this test](https://github.com/rust-lang/cargo/blob/1e768f712dfac8c093d46910a7582a3aec0e03de/tests/testsuite/cfg.rs) means that `foo` can see `bar 0.1.0` and `bar 0.2.0` depending on platform. This should not be a problem, but requires a deep level of disjointness analysis about platforms to be added to make it work.

alexcrichton replied:
For renaming though we want to give an error in that case, so renaming doesn't affect public/private, right? For platforms it's true that in theory we want to take the exhaustive list of any platforms and if none of them actually activate both dependencies we don't give an error, but it seems that for all practical purposes we'd want to give an error assuming all dependencies can be activated on at least one platform.

Eh2406 replied:
That is a reasonable position. (although different from what you suggested at https://github.com/rust-lang/cargo/issues/6129#issuecomment-465189082) It would mean that this existing stable behavior should not be allowed. Such a decision would require more discussion. I thought that we could have that discussion in a follow up PR if that is the behavior we decide on. 

This PR is the minimum functionality to unlock further work. I thought it would be more in kind for this PR to adopt the behavior of treating `cfg` and `rename` as having the "escape hatch" you suggested.

alexcrichton commented:
Do you think that we'll need to have fast paths for no public dependencies to preserve the current performance of the resolver? Or do the new code paths largely come out in the wash in profiling?

Eh2406 commented:
Sorry for the delay in replying, real life is being very difficult at the moment.

From what I have seen, this is not a hotspot. I have mostly been investigating cases of exponential blowout caused by this feature, and even there this inefficient code is not hot. I think with the use of Im-rs the general performance regression will will not be significant in absolute terms. That being said if you would like to check servo and x.py, it can't hurt.

Eh2406 commented:
Life has been kicking me this week, and will be for the foreseeable future. I found 2 mins to rebase this. I think let's get this landed and fix the problems in follow up PRs.

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -585,35 +590,80 @@ fn activate(
+                .chain(&Some(candidate_pid))
+                .cloned()
+                .collect();
+            for c in cs {
```
Could the documentation for this loop and the `cs` variable above be expanded to include what they're doing?

Eh2406 replied:
Added documentation for the `cs` variable. Now that I am typing a response I see that you wanted an explanation of the entire loop. BRB.

Eh2406 replied:
Comments updated more fully. Eventually this should probably be a dedicated data structure that keeps the related functionality together, and removes some of the duplication.

alexcrichton commented:
Ok I think if it's not showing up anywhere in profiles yet we should be good to add this, and we can always use bug reports on nightly to help guide us! I think with a few more comments (listed above) and the removal of the cfg/naming branches this should be good to land

alexcrichton commented on src/cargo/core/resolver/mod.rs:
```diff
@@ -487,6 +527,7 @@ fn activate_deps_loop(
                                 &parent,
                                 backtracked,
                                 &conflicting_activations,
+                                check_public_visible_dependencies,
```
Could this perhaps be stored in a `Context` or somewhere to avoid being passed around?

Eh2406 replied:
It could, and if you want to do it will be. Generally, Context storage things that change as resolution precedes. This will always have the same value. So I didn't see the need to store it repeatedly on the heap. Of course it's one bool, so it doesn't matter very much.

Would you like me to change it?

alexcrichton replied:
Yeah it's not the greatest spot for it, but I figure it's a bit better than passing and threading around to lots of places

Eh2406 replied:
Does that look better?

alexcrichton commented:
r=me with just one small stylistic nit about passing variables around

alexcrichton commented:
@bors: r+

Looks great to me!

bors commented:
:pushpin: Commit 12e474b03f587a82a7ea038fcadcb17c8056acfe has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 12e474b03f587a82a7ea038fcadcb17c8056acfe -->

bors commented:
:hourglass: Testing commit 12e474b03f587a82a7ea038fcadcb17c8056acfe with merge 0add586e3344f141b623ac1043b42321a3932de0...
<!-- homu: {"type":"BuildStarted","head_sha":"12e474b03f587a82a7ea038fcadcb17c8056acfe","merge_sha":"0add586e3344f141b623ac1043b42321a3932de0"} -->

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/cargo/builds/103281063)
<!-- homu: {"type":"BuildFailed","builder_url":"https://travis-ci.com/rust-lang/cargo/builds/103281063","builder_name":"checks-travis"} -->

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 12e474b03f587a82a7ea038fcadcb17c8056acfe with merge 2eb879bf7babc89dc3c32e47c547f0ef06902c34...
<!-- homu: {"type":"BuildStarted","head_sha":"12e474b03f587a82a7ea038fcadcb17c8056acfe","merge_sha":"2eb879bf7babc89dc3c32e47c547f0ef06902c34"} -->

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/cargo/builds/103289598)
<!-- homu: {"type":"BuildFailed","builder_url":"https://travis-ci.com/rust-lang/cargo/builds/103289598","builder_name":"checks-travis"} -->

Eh2406 commented:
@bors: retry

bors commented:
:hourglass: Testing commit 12e474b03f587a82a7ea038fcadcb17c8056acfe with merge cb0f7636192e986fe577d9c5da4a4afae7c468ad...
<!-- homu: {"type":"BuildStarted","head_sha":"12e474b03f587a82a7ea038fcadcb17c8056acfe","merge_sha":"cb0f7636192e986fe577d9c5da4a4afae7c468ad"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/cargo/builds/103305473), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/builds/22855262)
Approved by: alexcrichton
Pushing cb0f7636192e986fe577d9c5da4a4afae7c468ad to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/cargo/builds/22855262","checks-travis":"https://travis-ci.com/rust-lang/cargo/builds/103305473"},"merge_sha":"cb0f7636192e986fe577d9c5da4a4afae7c468ad"} -->
